### PR TITLE
Add device xFit features and tensor inference to XScan

### PR DESCRIPTION
**Pull request by tpn**
This PR lets XScan build classifier features directly from live xFit results, including a CuPy adapter that keeps the feature matrix on the GPU. The new `predict_tensors` API accepts image tensors and optional fit features already on one CUDA device and returns logits and probabilities on that device.

These APIs allow fitting and classification to share device data without an intermediate file or host-array conversion. Callers supply contiguous float32 tensors and manage stream ordering and shared storage lifetimes.

### CODE REVIEW

**Comment by tpn** [comment]
@coderabbitai full review

**Comment by tpn** [comment]
@coderabbitai full review

File: src/cuphoton/xscan/training.py [hunk 1]
```diff
@@ -194,6 +196,173 @@ def autocast_context(
...
+    except ImportError as exc:
+        raise ImportError(
+            "CuPy-to-Torch XScan handoff requires CuPy; run "
+            "'uv sync --extra gpu' for development or install "
+            "'cuphoton[gpu]'"
+        ) from exc
+    return cp
+
+
+def _cupy_to_torch(
```
Comment (melo-gonzo) [inline src/cuphoton/xscan/training.py]: should-fix: this is the only CuPy to torch bridge, and it is private, so the tests import _cupy_to_torch to compose DeviceXFitFeatures with predict_tensors. Users will need the same composition; expose the bridge (or a predict_device_features wrapper) so no private symbol is required.

File: src/cuphoton/xscan/xfit_features.py [hunk 2]
```diff
@@ -1070,6 +1142,519 @@ def _feature_row(
...
+        "converged": result.converged,
+        "degrees_of_freedom": result.degrees_of_freedom,
+        "valid_pixel_fraction": result.valid_pixel_fraction,
+        "fractional_null_improvement": result.fractional_null_improvement,
+        "delta_chi_square": result.delta_chi_square,
+        "reduced_chi_square": result.reduced_chi_square,
+        "uncertainty_valid": result.uncertainty_valid,
+        "standard_errors": result.standard_errors,
+        "covariance": result.covariance,
+        "residuals": result.residuals,
```
Comment (melo-gonzo) [inline src/cuphoton/xscan/xfit_features.py]: nit: the device source validation requires result.residuals (CuPy, compute dtype, shape) only to check image_shape, while the host adapter documents residuals as not required and not inspected. This forces callers to keep residuals resident on device; take image_shape from the config or parameters instead.

**Review by melo-gonzo — DISMISSED** [review]
Approving with nits. Column by column the device transform matches the canonical host row: same 17-column order, constants, clip ranges, fit_valid and uncertainty_valid gating, non-finite to zero, FP64 compute with one FP32 narrowing at the end. predict_tensors checks device, dtype, contiguity, and rank, rechecks model placement per call, runs under no_grad with per-module eval restore, and returns logits and probabilities on device; stream ordering is left to the caller and documented. The DLPack bridge hands the consumer stream to CuPy correctly. training.py and the tensor-inference tests are identical to GitLab main; xfit_features.py differs there only by the subnormal series that comes in #35, stacked on this. CI on the current run is fully green; the FAILURE entry is a superseded run cancelled by concurrency.

**Comment by tpn** [comment]
Addressed both approval nits in a49f354e and d36aca71:

- `cupy_to_torch` and `CupyTorchView` are public in `cuphoton.xscan.training`, so feature-to-inference composition no longer needs a private import. Existing private aliases remain compatible.
- Feature conversion uses the explicit `image_shape` and no longer inspects residuals. Callers can use `dataclasses.replace(result, residuals=None)` to release that storage; validation is unchanged when residuals are present.

The CPU suite passes (1,331 tests), along with 76 focused local CUDA tests. Multi-device cases remain skipped.

**Comment by tpn** [comment]
@coderabbitai full review

**Review by melo-gonzo — APPROVED** [review]
Re-approving after the rebase. a49f354 exposes cupy_to_torch as public with the private alias retained, and d36aca7 makes residuals optional on the device result so callers can release them once features are built, which removes the residual retention requirement I flagged. The base commit is unchanged in content. Both nits closed.

**Comment by tpn** [comment]
@coderabbitai full review

**Comment by tpn** [comment]
@coderabbitai full review
